### PR TITLE
refactor: migrated savings account application screen to compose

### LINKED ISSUE #2521: refactor: Migration of Savings Account Application Screen from XML to Compose
**Summary:**
Migration of Savings account application screen `ui/fragments/SavingsAccountApplicationFragment` from xml to compose.

### PR DESCRIPTION
**Pull request by AvneetSingh2001**
Fixes #2521 

Please Add Screenshots If there are any UI changes.

Before(XML Screen): 
https://github.com/openMF/mifos-mobile/assets/77102514/91c990b5-e8ff-47c8-865e-2893fe0461cd

After(Compose Screen): 
https://github.com/openMF/mifos-mobile/assets/77102514/f17bc7d2-84e8-4099-8a6a-53e355685db1

### CODE REVIEW

**Comment by therajanmaurya** [comment]
image
This screen does not follow the material guideline, please give 16dp padding from left and right.

**Comment by AvneetSingh2001** [comment]
hey, thats screenshot is actually the previous design in XML, that doesn’t have the padding. 

I have already given the padding in Compose design, you can check Compose screen recording here: 
https://github.com/openMF/mifos-mobile/assets/77102514/f17bc7d2-84e8-4099-8a6a-53e355685db1

File: app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationContent.kt [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+import org.mifos.mobile.models.templates.savings.SavingsAccountTemplate
+import org.mifos.mobile.utils.getTodayFormatted
+
+@Composable
+fun SavingsAccountApplicationContent(
+    existingProduct: String? = null,
+    savingsAccountTemplate: SavingsAccountTemplate? = null,
+    submit: (Int, Int) -> Unit
+) {
+    var selectProductId by remember { mutableIntStateOf(-1) }
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationContent.kt]: use `rememberSavable`

File: app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationFragment.kt [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+import javax.inject.Inject
+
+/*
+* Created by saksham on 30/June/2018
+*/
+@AndroidEntryPoint
+class SavingsAccountApplicationFragment : BaseFragment() {
+
+    @Inject
+    lateinit var preferencesHelper: PreferencesHelper
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationFragment.kt]: This also comes to data computations and It should be in ViewModel, move it there.

File: app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationFragment.kt [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+            return
+        }
+        payload.submittedOnDate = DateHelper.getSpecificFormat(
+            DateHelper.FORMAT_dd_MMMM_yyyy,
+            getTodayFormatted(),
+        )
+        viewModel.submitSavingsAccountApplication(payload)
+    }
+
+    private fun updateSavingAccount(productId: Int, clientId: Int) {
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationFragment.kt]: Move this to viewmodel

File: app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationFragment.kt [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        val payload = SavingsAccountUpdatePayload()
+        payload.clientId = clientId.toLong()
+        payload.productId = productId.toLong()
+        viewModel.updateSavingsAccount(viewModel.savingsWithAssociations?.id, payload)
+    }
+
+    private fun onSubmit(productId: Int, clientId: Int) {
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationFragment.kt]: move this to viewmodel

File: app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationFragment.kt [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+                                clientId = clientId
+                            )
+                        }
+                    )
+                }
+            }
+        }
+    }
+
+    private fun submitSavingsAccountApplication(productId: Int, clientId: Int) {
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationFragment.kt]: move this to viewmodel and add callback for toast so you can have access to context in compose and show toast.

submitSavingsAccountApplication(productId: Int, clientId: Int, showToast: (String) -> Unit)
now you can directly call this from view model to composable

File: app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationFragment.kt [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+    private fun onSubmit(productId: Int, clientId: Int) {
+        if (viewModel.savingsAccountState == SavingsAccountState.CREATE) {
+            submitSavingsAccountApplication(productId = productId, clientId = clientId)
+        } else {
+            updateSavingAccount(productId = productId, clientId = clientId)
+        }
+    }
+
+    private fun onRetry() {
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationFragment.kt]: Same move this to viewmodel

File: app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationScreen.kt [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+@Composable
+fun SavingsAccountApplicationScreen(
+    uiState: SavingsAccountApplicationUiState? = null,
+    savingsWithAssociations: SavingsWithAssociations? = null,
+    navigateBack: () -> Unit,
+    submit: (Int, Int) -> Unit,
+    retryConnection: () -> Unit,
+) {
+    var topBarTitleText by remember { mutableStateOf("") }
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationScreen.kt]: use remebersaveable

File: app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationScreen.kt [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+            is SavingsAccountApplicationUiState.ShowUserInterfaceSavingAccountApplication -> {
+                topBarTitleText = stringResource(id = R.string.apply_savings_account)
+                SavingsAccountApplicationContent(
+                    savingsAccountTemplate = uiState.template,
+                    submit = submit
+                )
+            }
+
+            is SavingsAccountApplicationUiState.ShowUserInterfaceSavingAccountUpdate -> {
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationScreen.kt]: we can merge these two ShowUserInterfaceSavingAccountUpdate and ShowUserInterfaceSavingAccountApplication by maintaining the field in ViewModel 
enum class RequestType {
 NEW_APPLICATION, 
 UPDATE
 }
 
 So initially we keep the value in ViewModel and set the parameter here accordingly.

File: app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationScreen.kt [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+                topBarTitleText = stringResource(id = R.string.update_savings_account)
+                SavingsAccountApplicationContent(
+                    existingProduct = savingsWithAssociations?.savingsProductName,
+                    savingsAccountTemplate = uiState.template,
+                    submit = submit
+                )
+            }
+
+            is SavingsAccountApplicationUiState.SavingsAccountApplicationSuccess -> {
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationScreen.kt]: Same way we can merge these two.

**Review by therajanmaurya — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationFragment.kt [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+import org.mifos.mobile.ui.fragments.base.BaseFragment
+import org.mifos.mobile.utils.Constants
+import org.mifos.mobile.utils.ParcelableAndSerializableUtils.getCheckedParcelable
+import org.mifos.mobile.utils.ParcelableAndSerializableUtils.getCheckedSerializable
+
+/*
+* Created by saksham on 30/June/2018
+*/
+@AndroidEntryPoint
+class SavingsAccountApplicationFragment : BaseFragment() {
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationFragment.kt:24]: Awesome Job! Now later on when we will implement compose navigation we only need to pick the compose and done.

**Review by therajanmaurya — APPROVED** [review]

File: app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationViewModel.kt [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+import org.mifos.mobile.utils.DateHelper
+import org.mifos.mobile.utils.getTodayFormatted
+import javax.inject.Inject
+
+@HiltViewModel
+class SavingsAccountApplicationViewModel @Inject constructor(private val savingsAccountRepositoryImp: SavingsAccountRepository) :
+    ViewModel() {
+
+    @Inject
+    lateinit var preferencesHelper: PreferencesHelper
```
Comment (therajanmaurya) [inline app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationViewModel.kt]: You don't neet to inject here, you can directly inject in constructor like 

@HiltViewModel
class SavingsAccountApplicationViewModel @Inject constructor(
private val savingsAccountRepositoryImp: SavingsAccountRepository, 
val preferencesHelper: PreferencesHelper
) : ViewModel() {

**Review by therajanmaurya — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationViewModel.kt (hunk 11 above)
Comment (AvneetSingh2001) [reply app/src/main/java/org/mifos/mobile/ui/savings_account_application/SavingsAccountApplicationViewModel.kt]: oh sorry, i missed that. Updated now

**Pull request merged**